### PR TITLE
Improve syntax for artifact.sync config [1/3]

### LINKED ISSUE #1180: File Sync format specification: maybe don't have user data in the key part
#1039 added a `key: value` map under `sync` in the skaffold format.
The key is a local glob pattern, the value is the destination directory.

Example:
```yaml
    sync:
      '*.py': .
```

Coming from k8s it is unusual to have user data as keys; it surprised me.

Digging into k8s design doc, we have:
https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#lists-of-named-subobjects-preferred-over-maps
There are whole discussions linked from there, notably about
- readability (mostly in case of subobjects though)
- merging definitions, which may or may not be of concert for skaffold

It lists exceptions to this though: "pure maps [...] as opposed to sets of subobjects".

I suppose if we later want to enrich the format with subobjects it would make things harder with the current format.

I am not (yet) familiar with all the skaffold format, I don't know if there are other cases like that, I couln't find any.

# Question

Shouldn't we forbit storing user data as keys?

# Suggestions

Either go full sub-object like in k8s:
```yaml
    sync:
    - localPattern: '*.py'
      destination: .
```
(better names are welcome)

Or maybe support lighter format with string:
```yaml
    sync:
    - '*.py:.'
```
But I'm not sure it's a good idea: it doesn't seem that much readable...

### PR DESCRIPTION
**Pull request by corneliusweig**
Currently, the pipeline config is a map of local glob pattern to destination directory. This scheme has been extended with some magic sequences, making it difficult to understand. This commit converts the sync map into a list of sync rules. All sync rules are considered to determine the destination paths.
    
This is the first step of sync improvements as detailed out in the design proposal #1844.

Extracted from #1812
Close #1180

CC design shepherd @tejal29

### CODE REVIEW

**Comment by corneliusweig** [comment]
@tejal29 As you were shepherd of the design proposal, you might also want to take a look at this :)

**Comment by tejal29** [comment]
Thanks @corneliusweig I was waiting for Design proposal to merge in.

**Comment by tejal29** [comment]
Thanks @corneliusweig I was waiting for Design proposal to merge in.

**Comment by corneliusweig** [comment]
Right, sorry for being impatient :bowing_man:

**Comment by corneliusweig** [comment]
@tejal29 Now that the sync design proposal is accepted, it's time to move on with the sync config. I therefore rebased this changeset to `master` changes and updated the docs.

**Comment by corneliusweig** [comment]
@tejal29 do you see a chance this gets merged into v0.29.0?

**Comment by corneliusweig** [comment]
@nkubala You mentioned in the bi-weekly that people complained about the sync experience. I think a clearer sync config already helps a lot. Any chance this PR can be prioritized?

**Comment by tejal29** [comment]
@corneliusweig Sorry we ignored this one for long!
I will take a look as soon as get to work today.

Thank you for your patience.

**Comment by corneliusweig** [comment]
> I will take a look as soon as get to work today.

Hey @tejal29 , thanks for giving this such high priority! Very much appreciated!

File: docs/content/en/samples/filesync/filesync.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  artifacts:
+    - image: gcr.io/k8s-skaffold/node-example
+      context: node
+      sync:
+        manual:
+          - src: '.filebaserc'
+            dest: /etc
+          - src: 'static-html/*.html'
+            dest: static
+          - src: '**/*.png'
```
[the remark this answers is not in the document]
Comment (corneliusweig) [reply docs/content/en/samples/filesync/filesync.yaml]: I've added some comments. Please have a look if this is what you intended. I'm not sure whether I like those comments, though. After all, each rule is explained further below. But I trust you have a better feeling for what's appropriate here.

File: pkg/skaffold/schema/validation/validation_test.go [hunk 2]
```diff
@@ -348,6 +348,104 @@ func TestValidateNetworkMode(t *testing.T) {
...
+				{
+					Sync: &latest.Sync{Manual: []*latest.SyncRule{
+						{
+							Src:  "src/**/*.js",
+							Dest: ".",
+						},
+						{
+							Src:   "src/**/*.js",
+							Dest:  ".",
+							Strip: "src/",
```
[the remark this answers is not in the document]
Comment (corneliusweig) [reply pkg/skaffold/schema/validation/validation_test.go]: done

**Comment by corneliusweig** [comment]
@tejal29 nits done. PTAL

File: docs/content/en/docs/how-tos/filesync/_index.md [hunk 3]
```diff
@@ -10,40 +10,38 @@ This page discusses how to set up file sync for files that don't require full re
...
+The file copying is enabled by adding a `sync` section with _sync rules_ to the `artifact` in the `skaffold.yaml`.
+Under the hood, Skaffold creates a tar file with changed files that match the sync rules.
+This tar file is sent to and extracted on the corresponding containers. 
+
+### Manual sync mode
+
+A manual sync rule must specify the `src` and `dest` field.
+The `src` field is a glob pattern to match files relative to the artifact _context_ directory, which may contain `**` to match nested files.
+The `dest` field is the destination location in the container.
+It may be absolute or relative, in which case files are put below the container's `WORKDIR`.
```
Comment (priyawadhwa) [inline docs/content/en/docs/how-tos/filesync/_index.md]: Just for clarity, I would rephrase this as `Destination paths may be absolute or relative. If the destination is a relative path, an absolute path will be inferred by prepending the path with the container's WORKDIR`

File: pkg/skaffold/schema/v1beta9/upgrade.go [hunk 4]
```diff
@@ -17,18 +17,26 @@ limitations under the License.
...
+
 	next "github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/util"
 	pkgutil "github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
 	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	incompatibleSyncWarning = `the semantics of the sync rules has changed, the folder structure is not flattened anymore but preserved, the likely impacted patterns in your skaffold yaml are: %s`
```
Comment (priyawadhwa) [inline pkg/skaffold/schema/v1beta9/upgrade.go]: Might be nice to provide a link to the docs here so people can easily find the new sync rules

**Comment by corneliusweig** [comment]
@priyawadhwa Thanks for your valuable input! PTAL

**Comment by tejal29** [comment]
Thanks @corneliusweig, i have brought this to this week's release master's attention :)
They will merge it in. 
From my side #LGTM

**Comment by corneliusweig** [comment]
@tejal29 Unless I completely misunderstood you, you hit the wrong button ;)

**Comment by balopat** [comment]
haha, I think it was accidental :D

**Comment by corneliusweig** [comment]
@balopat Oh, it was a pleasure (most of the time :)

About `syncMap`, do you think about a local type alias in the `sync` package, or more on the global scope alongside `interface Builder`? Anyway, you're right, there really should be a type alias.

**Comment by balopat** [comment]
I'm rerunning kokoro - this seems to be a flake...

```
time="2019-05-09T22:32:42Z" level=info msg="Ran in 6.011332858s"
--- FAIL: TestGetStateRPC (7.12s)
    rpc_test.go:286: waiting for connection...
    rpc_test.go:246: skaffold build or deploy not complete. state: buildState:<artifacts:<key:"gcr.io/k8s-skaffold/test-dev" value:"In Progress" > > deployState:<status:"Not Started" >
```

if it is, I'll open a bug.

**Comment by balopat** [comment]
> @balopat Oh, it was a pleasure (most of the time :)
> 
> About `syncMap`, do you think about a local type alias in the `sync` package, or more on the global scope alongside `interface Builder`? Anyway, you're right, there really should be a type alias.

I'd keep it in `sync` - seems like a local concept for now. we should pull it up to a wider scope only if required.

**Comment by corneliusweig** [comment]
What a thriller, these integration tests...

**Comment by balopat** [comment]
https://github.com/GoogleContainerTools/skaffold/issues/2100 for the flake
